src/main/java/org/truffleruby/core/thread/RubyThread.java

```diff
@@ -76,6 +79,7 @@ public RubyThread(
         this.finishedLatch = new CountDownLatch(1);
         this.threadLocalVariables = HashOperations.newEmptyHash(context);
         this.recursiveObjects = HashOperations.newEmptyHash(context);
+        this.recursiveObjectSet = EconomicSet.create(Equivalence.IDENTITY_WITH_SYSTEM_HASHCODE);
```

Comment:
This seems incorrect if it can contain things like `java.lang.Integer` which are `!=` in Java but `.equal?` in Ruby.
`IDENTITY` is probably fine as we keep hashCode() as the original one for RubyDynamicObject, etc.

Reply:
recursiveObjectSet should only contain things that can contain things (i.e arrays, hashes, structs).

Reply:
Good point, single `detect_recursion` is only on Array/Hash/Enumerator, and also in `foreign_inspect`, where it's probably OK to compare by Java identity for the purpose of detecting recursion.

Could you `assert object instanceof RubyDynamicObject || RubyGuards.isForeignObject(object)` in `detect_recursion`?